webui/react/src/stores/roles.tsx

```diff
@@ -1,22 +1,31 @@
 import { listRoles } from 'services/api';
+import { ListRolesParams } from 'services/types';
 import { UserRole } from 'types';
 import handleError from 'utils/error';
 import { Loadable, Loaded, NotLoaded } from 'utils/loadable';
-import { observable, useObservable, WritableObservable } from 'utils/observable';
+import { observable, WritableObservable } from 'utils/observable';
 
-export class RolesStore {
-  static #roles: WritableObservable<Loadable<UserRole[]>> = observable(NotLoaded);
+class RoleStore {
+  #roles: WritableObservable<Loadable<UserRole[]>> = observable(NotLoaded);
 
-  static fetchRoles = async (canceler: AbortController): Promise<void> => {
-    try {
-      const response = await listRoles({ limit: 0 }, { signal: canceler.signal });
-      this.#roles.set(Loaded(response));
-    } catch (e) {
-      handleError(e);
-    }
-  };
+  public readonly roles = this.#roles.readOnly();
```

Comment:
question: why not using getter?

Reply:
Are you thinking something like:
```
get roles() {
  return this.#roles.readOnly();
}
```
or did you have something else in mind?

Reply:
yes thats what i was thinking
also, nice to add return type 

```tsx
get roles(): Read only type {
  return this.#roles.readOnly();
}
```

Reply:
I agree that getter / setter are redundant, but the main reason why i prefer getter / setter is that its the recommended way for OPP in general. the second reason is it's type safer [typescript playground](https://www.typescriptlang.org/play?removeComments=true&target=99&ts=4.9.4#code/MYGwhgzhAEDCYBdoG8CwAoaXoAcCmAThAPYB2YIAlggJ4BcASnmACZkg0A8TwxBLnCAgKVSAcwA00ISPEA+OVgC8KDNmxg60AORhta7AF8A3BgNYCzNqQ7QAbmBFgARiDxaefATNGTpw30VoFTRMdWhNHT1zaBMzMKwxPCQkhAAFQhJyKloACgBKDyt2Lk9+QQDxKR95RVDw6EsEAFcCUmgEAAtKCAA6fCIyCmoaGMN49V5SGWbgBD4ClHH0ZYwpoWhgRGDoUjwAdzhEAtN0DAB6c4jgYDwoe0dKFzdofepO4maU5IRCNcR+pkhjkaABtXTaAC6O10EBY2mgl02YHaYDsxEoLFwBGIA1oESglDEpAAtnhSAh-ghAYNsiMdshItoWAAzOEs7SGRFXLYUiLozEPJyuPAEiCiMkU+JbakOYVucF6aEqbTs-ToGW9OVPEUMlhaZlwznxeJIsA3O4wbXPUVdT5iTrQVK-AhrMgbMA7TWpDK04a0C48lH8jFY4CdFFiXwPEDNO4YT0hJnw5ZgRVQmGQeEYIA)

Reply:
I get what you are saying now. IMHO, because with the stores, we are dealing with read-only observables specifically, it's not possibly to modify the observable value as a consumer of the instance member variable.

In the case of objects and setting a value for a key or any other complex objects, immutables and the likes might offer better protection from accidentally mutating them.

Reply:
> but the main reason why i prefer getter / setter is that its the recommended way for OPP in general

I can understand this sentiment. JS was pretty lawless for a while but with its improvements and the TS layer there are safeguards we benefit from. From OOP's perspective it's purely a syntax sugar on top of Object.create and prototypes, so IMO it's reasonable to not follow OOP to the most strict degree. The benefits of getters and setters, I'm more in the camp of them not being very helpful, especially the ones that are simply one line returns or updates. My proposal is to maybe also discuss this with the team. Wdy think?

Reply:
im fine with the way without getter / setter if you are confident because you know the store implementation more than me, so i think team discussion is optional unless someone else have opinions,